### PR TITLE
Refactor code checker to use XML AST representation

### PR DESCRIPTION
**Pull request by LukasReschke**
While trying to traverse through the AST of a file I noticed that there is no way using a visitor to find the actual context of a call (i.e. it is not possible to traverse up in the AST), I do require this since I actually want to also catch things like `(string) $_GET['foo']` whereas there is currently no reliable way to get this without a lot of black magic.

I know that XPath and XML are certainly not everybody's best friend. However it gives us way more flexibility and since I splitted the single checks in their own files the code should also be way easier to maintain for the future.

Once this is merged I can come up with further security relevant checks.

@DeepDiver1975 @MorrisJobke Please review

### CODE REVIEW

File: lib/private/app/codechecker/codechecker.php [hunk 1]
```diff
@@ -47,11 +53,15 @@ class CodeChecker extends BasicEmitter {
...
+	/** @var XML */
+	private $serializer;
+
 	/** @var string[] */
 	private $blackListedClassNames;
 
 	public function __construct() {
 		$this->parser = new Parser(new Lexer);
+		$this->serializer = new XML();
 		$this->blackListedClassNames = [
```
Comment (LukasReschke) [inline lib/private/app/codechecker/codechecker.php]: Shall we still use this or use my heuristic which does just check for `OC_`?

File: lib/private/app/codechecker/check.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		$this->errors[] = $error;
+	}
+
+	/**
+	 * @return Error[]
+	 */
+	public function getErrors() {
+		return $this->errors;
+	}
+}
```
Comment (LukasReschke) [inline lib/private/app/codechecker/check.php]: Newline at EoF – will add at next rebase / commit

File: core/command/app/checkcode.php [hunk 3]
```diff
@@ -41,20 +42,28 @@ protected function configure() {
...
 		$codeChecker->listen('CodeChecker', 'analyseFileBegin', function($params) use ($output) {
 			$output->writeln("<info>Analysing {$params}</info>");
 		});
 		$codeChecker->listen('CodeChecker', 'analyseFileFinished', function($params) use ($output) {
+			/** @var Error[] $params */
+
 			$count = count($params);
 			$output->writeln(" {$count} errors");
-			usort($params, function($a, $b) {
+			// FIXME: Make work with error class
```
Comment (LukasReschke) [inline core/command/app/checkcode.php]: Any ideas? – But I actually think it should already obey the line count since it is an AST?

File: tests/lib/app/codechecker/codechecker.php [hunk 4]
```diff
@@ -20,12 +21,13 @@ class CodeChecker extends TestCase {
...
-		$checker = new OC\App\CodeChecker();
+		$checker = new OC\App\CodeChecker\CodeChecker();
+		/** @var Error[] $errors */
 		$errors = $checker->analyseFile(OC::$SERVERROOT . "/tests/data/app/code-checker/$fileToVerify");
 
 		$this->assertEquals(1, count($errors));
-		$this->assertEquals($expectedErrorCode, $errors[0]['errorCode']);
-		$this->assertEquals($expectedErrorToken, $errors[0]['disallowedToken']);
+		$this->assertEquals($expectedErrorCode, $errors[0]->getCode());
+		$this->assertEquals($expectedErrorToken, $errors[0]->getDisallowedToken());
```
Comment (LukasReschke) [inline tests/lib/app/codechecker/codechecker.php]: We probably should also add the lines and the message to the unit tests.

File: lib/private/app/codechecker/checks/privatecode/implementsprivateclasses.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	 */
+	public function update(SplSubject $subject) {
+		/** @var $subject Check */
+		$xpath = $subject->xmlAst->xpath('//node:Stmt_Class/subNode:implements/scalar:array/node:Name');
+
+		if(!empty($xpath)) {
+			foreach($xpath as $result) {
+				// For each element get the name
+				$name = (string)$result->xpath('./subNode:parts/scalar:array/scalar:string')[0];
+				if(substr(strtoupper($name), 0, 3) === "OC_") {
```
Comment (DeepDiver1975) [inline lib/private/app/codechecker/checks/privatecode/implementsprivateclasses.php]: this completly changes behavior as it kills any use of OC_* classes - while generally speaking this is the "right-thing-to-do" - I have to disagree at this point in time - I'd honestly be happy if all apps just fulfill the current existing limitations with respect to the backlisted classes - for 8.2 we can switch on a more general pattern.

Please keep in mind that the code checker will disallow app installation - even a tough barrier with the current limit with respect to existing blacklisted classes.

Long story short - shall we postpone this to 8.2? THX @LukasReschke

Comment (LukasReschke) [inline lib/private/app/codechecker/checks/privatecode/implementsprivateclasses.php]: I'm fine with 8.2 as well hopong that not too much things will change here to get this in. - Though adjusting the behaviour wouldn't take too much time as it would simply be a in_array search.

That said I'd still appreciate a review of this as I plan to maintain this branch further to have some generic security checks. I really don't want to write a lot of stuff and then have to rewrite it :smiley:


Ref https://github.com/owncloud/core/pull/16076#diff-for-comment-29689064

Comment (LukasReschke) [inline lib/private/app/codechecker/checks/privatecode/implementsprivateclasses.php]: => Completely your call. Go with what you prefer. - You are the master of core and code checking ;-)

**Comment by MorrisJobke** [comment]
> Please keep in mind that the code checker will disallow app installation - even a tough barrier with the current limit with respect to existing blacklisted classes.

Really? Then the `==` and `!=` are also blocking an app installation? Then I would vote for a revert - otherwise we have a broken universe

**Comment by LukasReschke** [comment]
> Really? Then the == and != are also blocking an app installation? Then I would vote for a revert - otherwise we have a broken universe

Well, pechi-pechi, one shall have followed our [coding guidelines](https://doc.owncloud.org/server/8.0/developer_manual/general/codingguidelines.html#operators) from the start on :see_no_evil:

That said, should be easy to fix in all the apps as it is just a "search and replace" that often works quite fine.

**Comment by MorrisJobke** [comment]
> That said, should be easy to fix in all the apps as it is just a "search and replace" that often works quite fine.

-.- No, please don't do that. It has a huge potential to brick everything and you have no idea why.

**Comment by LukasReschke** [comment]
> -.- No, please don't do that. It has a huge potential to brick everything and you have no idea why.

I'm not saying that *I* do that. I'm saying that our usual app dev can do that and check the results.

File: lib/private/app/codechecker/checks/privatecode/implementsprivateclasses.php (hunk 5 above)
Comment (LukasReschke) [inline lib/private/app/codechecker/checks/privatecode/implementsprivateclasses.php]: https://github.com/owncloud/core/commit/29b6566d3461f850ce3e25c3270306302b620c20 :see_no_evil:

**Comment by LukasReschke** [comment]
@DeepDiver1975 @MorrisJobke @nickvergessen If we really want to make things like `==` only a soft-fail I'd propose that we go with this and just add a new error type `SoftFail` which will not have impact on whether the app can be installed. However, we will require following our code style (which includes this) for applications that want an approved tag in the appstore etc… – Way better than to hack the existing code…

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

**Comment by LukasReschke** [comment]
We now have a hard-fail class that will fail hard when installing app and a soft fail class that only throws a warning in OCC. – Also proper unit tests on the error object.

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

File: lib/private/app/codechecker/checks/security/comparisonoperators.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+	/**
+	 * @param SplSubject $subject
+	 */
+	public function update(SplSubject $subject) {
+		/** @var $subject Check */
+		$this->checkEqual($subject);
+		$this->checkNotEqual($subject);
+	}
+}
```
Comment (LukasReschke) [inline lib/private/app/codechecker/checks/security/comparisonoperators.php]: TODO: Add newline at next rebase.

**Comment by PVince81** [comment]
@LukasReschke when using visitors another possibility is to have two visitor classes go through the code. The first one gathers the necessary information and stores it in a catalog and the second visitor can then use it. But doing it that way might me more hassle.

**Comment by Xenopathic** [comment]
(In the future) We should introduce a check for a `new` being invoked with a variable, e.g. `new $class()`. This has the tendency to introduce security bugs...

File: lib/private/app/codechecker/checks/security/comparisonoperators.php (hunk 6 above)
Comment (PVince81) [inline lib/private/app/codechecker/checks/security/comparisonoperators.php]: The time has come it seems :smile:
